python/pyarrow/_compute.pyx

```diff
@@ -2047,6 +2047,60 @@ class RandomOptions(_RandomOptions):
         self._set_options(length, initializer)
 
 
+cdef class _RankOptions(FunctionOptions):
+
+    _tiebreaker_map = {
+        "Min": CRankOptionsTiebreaker_MIN,
+        "Max": CRankOptionsTiebreaker_MAX,
+        "First": CRankOptionsTiebreaker_FIRST,
+        "Dense": CRankOptionsTiebreaker_DENSE,
+    }
+
+    def _set_options(self, sort_keys, null_placement, tiebreaker):
+        cdef vector[CSortKey] c_sort_keys
+        for name, order in sort_keys:
+            c_sort_keys.push_back(
+                CSortKey(tobytes(name), unwrap_sort_order(order))
+            )
+        try:
+            self.wrapped.reset(
+                new CRankOptions(c_sort_keys, unwrap_null_placement(null_placement), self._tiebreaker_map[tiebreaker])
+            )
+        except KeyError:
+            _raise_invalid_function_option(tiebreaker, "tiebreaker")
+
+
+class RankOptions(_RankOptions):
+    """
+    Options for the `rank` function.
+
+    Parameters
+    ----------
+    sort_keys : sequence of (name, order) tuples
```

Comment:
@pitrou do you mean something like this change?
```
$ git diff
diff --git a/python/pyarrow/_compute.pyx b/python/pyarrow/_compute.pyx
index 6df6637..a4830d9 100644
--- a/python/pyarrow/_compute.pyx
+++ b/python/pyarrow/_compute.pyx
@@ -2076,7 +2076,7 @@ class RankOptions(_RankOptions):
 
     Parameters
     ----------
-    sort_keys : sequence of (name, order) tuples
+    sort_keys : sequence of (name, order) tuples, optional
         Names of field/column keys to sort the input on,
         along with the order each field/column is sorted in.
         Accepted values for `order` are "ascending", "descending".
@@ -2097,7 +2097,7 @@ class RankOptions(_RankOptions):
                    number of distinct values in the input.
     """
 
-    def __init__(self, sort_keys, *, null_placement="at_end", tiebreaker="First"):
+    def __init__(self,  *, sort_keys=(), null_placement="at_end", tiebreaker="First"):
         self._set_options(sort_keys, null_placement, tiebreaker)
 
 
diff --git a/python/pyarrow/tests/test_compute.py b/python/pyarrow/tests/test_compute.py
index 4b7aa3e..aaefefc 100644
--- a/python/pyarrow/tests/test_compute.py
+++ b/python/pyarrow/tests/test_compute.py
@@ -2764,8 +2764,7 @@ def test_rank_options_tiebreaker(tiebreaker, expected_values):
 
 def test_rank_options():
     arr = pa.array([1.2, 0.0, 5.3, None, 5.3, None, 0.0])
-    rank_options = pc.RankOptions(sort_keys=[("b", "ascending")])
-    result = pc.rank(arr, options=rank_options)
+    result = pc.rank(arr, options=pc.RankOptions())
     expected = pa.array([3, 1, 4, 6, 5, 7, 2], type=pa.uint64())
     assert result.equals(expected)
```

Reply:
Although for `sort_indices` we have a separate array-specific kernel with a separate ArraySortOptions (which is simpler, not requiring a column name), in addition to the general SortOptions. So that's a bit different here.

For the `sort_keys` keyword in top_k / bottom_k kernel, we have some special casing as well:

https://github.com/apache/arrow/blob/62979d249952775ebd06c814b16d58b87e1cbc88/python/pyarrow/compute.py#L544-L547

(the "descending" is hardcoded here because it is for `top_k`, but the relevant part is that is injects a "dummy" name for the column in case of array input)

Reply:
For the record, we probably shouldn't be using "dummy" as that can be an actual column name. An empty string might be better.

Reply:
@raulcd You should even be able to omit the `options` in the call to `pc.rank()`.

Reply:
I have pushed a new commit with some changes. I've added a test to ensure we can call `pc.rank()` without options and I've just applied the diff I proposed previously. I am not entirely sure I understand the proposal around injecting a fake ordering and what are the benefits vs the current implementation. With the current implementation we can use the following signatures:
```
    pc.rank(arr)
    pc.rank(arr, options=pc.RankOptions())
    pc.rank(arr, options=pc.RankOptions(tiebreaker="dense"))
    pc.rank(arr, options=pc.RankOptions(null_placement="at_start"))
    rank_options = pc.RankOptions(sort_keys=[("b", "ascending")],
                                  null_placement="at_end",
                                  tiebreaker=tiebreaker)
    pc.rank(arr, options=rank_options)
```

Reply:
> Although for `sort_indices` we have a separate array-specific kernel with a separate ArraySortOptions (which is simpler, not requiring a column name), in addition to the general SortOptions. So that's a bit different here.

If it makes it easier I can add this back into the C++ implementation

Reply:
> If it makes it easier I can add this back into the C++ implementation.

I would rather not. The array-specific kernel for sorting is not a very user-friendly API pattern. Let's just make `RankOptions` usable enough in Python :-)